File: demo/go/sub_unsub/sub_unsub.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
	"time"

	common "github.com/MuggleWei/babel-trader/src/babeltrader-common-go"
)

func TopicGet(addr string) ([]byte, error) {
	url := "http://" + addr + "/topic/get"
	return common.HttpRequest(http.DefaultClient, "GET", "", url, nil)
}

func TopicSub(addr string, msg *common.MessageSubUnsub) ([]byte, error) {
	url := "http://" + addr + "/topic/sub"
	bytes, err := json.Marshal(msg)
	if err != nil {
		return nil, err
	}
	return common.HttpRequest(http.DefaultClient, "Post", string(bytes), url, nil)
}

func TopicUnsub(addr string, msg *common.MessageSubUnsub) ([]byte, error) {
	url := "http://" + addr + "/topic/unsub"
	bytes, err := json.Marshal(msg)
	if err != nil {
		return nil, err
	}
	return common.HttpRequest(http.DefaultClient, "Post", string(bytes), url, nil)
}

func main() {
	// ctp demo
	// addr := "127.0.0.1:6001"

	// xtp demo
	addr := "127.0.0.1:6002"
	quote := common.MessageSubUnsub{
		Market:   "xtp",
		Exchange: "SSE",
		Type:     "spot",
		Symbol:   "601857",
	}

	// get all topics
	rsp, err := TopicGet(addr)
	if err != nil {
		log.Printf("[Error] failed in /topic/get: %s", err.Error())
		os.Exit(-1)
	}
	log.Printf("%s", string(rsp))

	// sub topic
	rsp, err = TopicSub(addr, &quote)
	if err != nil {
		log.Printf("[Error] failed in /topic/sub: %s", err.Error())
		os.Exit(-1)
	}
	log.Printf("%s", string(rsp))

	time.Sleep(time.Second * 3)

	rsp, err = TopicGet(addr)
	if err != nil {
		log.Printf("[Error] failed in /topic/get: %s", err.Error())
		os.Exit(-1)
	}
	log.Printf("%s", string(rsp))

	// unsub topic
	rsp, err = TopicUnsub(addr, &quote)
	if err != nil {
		log.Printf("[Error] failed in /topic/sub: %s", err.Error())
		os.Exit(-1)
	}
	log.Printf("%s", string(rsp))

	time.Sleep(time.Second * 3)

	rsp, err = TopicGet(addr)
	if err != nil {
		log.Printf("[Error] failed in /topic/get: %s", err.Error())
		os.Exit(-1)
	}
	log.Printf("%s", string(rsp))
}

File: demo/python/sub_unsub.py
import time

import requests

# ctp demo
addr = "127.0.0.1:6001"
sub_msg = {
    "market": "ctp",
    "type": "future",
    "symbol": "rb",
    "contract": "1905"
}

# # xtp demo
# addr = "127.0.0.1:6002"
# sub_msg = {
#     "market": "xtp",
#     "exchange": "SSE",
#     "type": "spot",
#     "symbol": "601857",
# }


def output_subed_topic():
    r = requests.post("http://" + addr + "/topic/get", json={})
    print(r.text)


def sub():
    r = requests.post("http://" + addr + "/topic/sub", json=sub_msg)
    print(r.text)


def unsub():
    r = requests.post("http://" + addr + "/topic/unsub", json=sub_msg)
    print(r.text)

if __name__ == '__main__':
    output_subed_topic()

    sub()
    time.sleep(3)
    output_subed_topic()

    unsub()
    time.sleep(3)
    output_subed_topic()

